staging/src/k8s.io/apiserver/pkg/endpoints/openapi/cacher.go

```diff
@@ -0,0 +1,234 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package openapi
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	golangproto "github.com/golang/protobuf/proto"
+	openapi_v2 "github.com/googleapis/gnostic/OpenAPIv2"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilresponsewriter "k8s.io/apiserver/pkg/util/responsewriter"
+	"k8s.io/kube-openapi/pkg/util/proto"
+)
+
+// NewSchemaCacher creates a new SchemaCacher
+func NewSchemaCacher(handler http.Handler, path string) *SchemaCacher {
+	return &SchemaCacher{
+		openAPIHandler: handler,
+		openAPIPath:    path,
+	}
+}
+
+// SchemaCacher fetches the openapi schema once and then caches it in memory
+type SchemaCacher struct {
+	rwMutex       sync.RWMutex
+	openAPISchema Resources
+	lastEtag      string
+
+	// These two fields will never change once the cacher is set up
+	// They sepcify where this cacher will try to download the schema from
+	openAPIHandler http.Handler
+	openAPIPath    string
+}
+
+var _ Getter = &SchemaCacher{}
+
+// Get implements Getter
+func (s *SchemaCacher) Get() (Resources, error) {
+	err := s.updateSchema()
+	if err != nil {
+		return nil, err
+	}
+
+	s.rwMutex.RLock()
+	defer s.rwMutex.RUnlock()
+	return s.openAPISchema, nil
+}
+
+// updateSchema parses the OpenAPI schema if it has changed since the last
+// time updateSchema was called, and cache it.
+func (s *SchemaCacher) updateSchema() error {
+	lastEtag := s.getLastEtag()
```

Comment:
you mean I should just say this? I thought you needed to use defer unlock() immediately after every lock()
```
s.rwMutex.RLock()
lastEtag := s.lastEtag
s.rwMutex.RUnlock()
```

Reply:
Yes

Reply:
drive by: I actually think it's fine and actually clearer the way Jenny has it and should probably stay this way until it's profiled & there's a demonstrable performance problem.

Reply:
I'll explain why I said that, and let you decide @jennybuckley :-)

Methods that grab mutex usually require somewhat more context (as in, you need to know that this is going to grab the mutex, so you can't call it from a function that already holds the mutex). I don't think that extra complexity is required here since: 1. it's only doing one very simple thing, 2. i think it's clearer from the caller that you're grabbing the mutex.

But no strong feeling.